### PR TITLE
AdaGrad classification version added; Data set from CS 446 HW 3 added

### PR DESCRIPTION
**Pull request by yimingjiang**
@danyaljj @christos-c Please review this PR.

* `AdaGradCL` classification added
* `AlgoDataSet`: CS 446 Homework 3 data set
* `AlgoData`, `AlgoParser`: example data structure, parser
* `AdaGradCLClassifier` added

With data set of 500 features and 50000 examples (40000 for training, 10000 for testing), AdaGradCL has 99.61% accuracy with only 10 iterations.

Thanks!

### CODE REVIEW

**Comment by danyaljj** [comment]
Do you think there is a way to have a single AdaGrad for both regression and classification?

**Comment by yimingjiang** [comment]
@danyaljj 
That is not quite possible, given the current inheritance architecture.

They both share some same interface, from `Learner`, such as `classify`. And, the necessary components for regression and classification are dramatically different, as the implementation shows. Classification and regression have their own unique set of functions to work properly.

**Comment by yimingjiang** [comment]
@danyaljj 
FYI:

`scikit-learn`, a popular python ML lib, provides [`SGDClassifier`](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html#sklearn.linear_model.SGDClassifier) and [`SGDRegressor`](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDRegressor.html#sklearn.linear_model.SGDRegressor).

**Comment by yimingjiang** [comment]
@danyaljj @christos-c Made a minor update. Added a debug utility class to help with debugging. I found it quite useful.

Please review the changes.

File: lbjava/src/test/java/edu/illinois/cs/cogcomp/lbjava/algo_test/main/AdaGradCLMain.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
+/**
```
Comment (danyaljj) [inline lbjava/src/test/java/edu/illinois/cs/cogcomp/lbjava/algo_test/main/AdaGradCLMain.java]: Do we need this @yimingjiang ?

Comment (yimingjiang) [inline lbjava/src/test/java/edu/illinois/cs/cogcomp/lbjava/algo_test/main/AdaGradCLMain.java]: We will unify the AdaGrad and SGD, with classification and regression versions. I will work on this once MLP works.

Comment (danyaljj) [inline lbjava/src/test/java/edu/illinois/cs/cogcomp/lbjava/algo_test/main/AdaGradCLMain.java]: Checking the folder here: https://github.com/yimingjiang/lbjava/tree/adagrad-cl/lbjava/src/test/java/edu/illinois/cs/cogcomp/lbjava/algo_test 
If the tests that you are adding are not sensitive to failures (like by assertions), we won't notice if something fails. We probably need to modify them so that they fail upon running through CI.

Comment (yimingjiang) [inline lbjava/src/test/java/edu/illinois/cs/cogcomp/lbjava/algo_test/main/AdaGradCLMain.java]: Sure. I will keep this in mind.

**Comment by christos-c** [comment]
@yimingjiang what is the status of this? Do you want to update it?

**Comment by yimingjiang** [comment]
@christos-c I will make a new PR soon. This PR is outdated. I will close it.
